invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java

```diff
@@ -160,54 +163,84 @@ private void onCompletionCall(final boolean isSuccess,
             // We clone the original invoice so plugins don't remove/add items
             final Invoice clonedInvoice = (Invoice) originalInvoice.clone();
+            // Invoice items returned by the plugin: some can be new items, some can be updated items (on the same or new invoice)
             final List<InvoiceItem> additionalInvoiceItemsForPlugin = invoicePlugin.getAdditionalInvoiceItems(clonedInvoice, isDryRun, properties, callContext);
+            if (additionalInvoiceItemsForPlugin == null || additionalInvoiceItemsForPlugin.isEmpty()) {
+                continue;
+            }
 
-            if (additionalInvoiceItemsForPlugin != null && !additionalInvoiceItemsForPlugin.isEmpty()) {
-                final Collection<InvoiceItem> additionalInvoiceItems = new LinkedList<InvoiceItem>();
-                for (final InvoiceItem additionalInvoiceItem : additionalInvoiceItemsForPlugin) {
-                    final InvoiceItem sanitizedInvoiceItem = validateAndSanitizeInvoiceItemFromPlugin(originalInvoice, additionalInvoiceItem, invoicePlugin);
-                    additionalInvoiceItems.add(sanitizedInvoiceItem);
-                }
-                invoiceUpdated = updateOriginalInvoiceWithPluginInvoiceItems(originalInvoice, additionalInvoiceItems) || invoiceUpdated;
+            for (final InvoiceItem additionalInvoiceItem : additionalInvoiceItemsForPlugin) {
+                final InvoiceItem sanitizedInvoiceItem = validateAndSanitizeInvoiceItemFromPlugin(originalInvoice, additionalInvoiceItem, invoicePlugin);
+                updatePerInvoiceInvoiceItems(perInvoiceInvoiceItems, sanitizedInvoiceItem);
             }
         }
 
-        return invoiceUpdated;
-    }
+        final List<DefaultInvoice> updatedInvoices = new LinkedList<DefaultInvoice>();
+        for (final UUID invoiceId : perInvoiceInvoiceItems.keySet()) {
+            final Collection<InvoiceItem> updatedInvoiceItems = perInvoiceInvoiceItems.get(invoiceId);
+            if (updatedInvoiceItems.isEmpty()) {
+                continue;
+            }
 
-    private boolean updateOriginalInvoiceWithPluginInvoiceItems(final DefaultInvoice originalInvoice, final Collection<InvoiceItem> additionalInvoiceItems) {
-        if (additionalInvoiceItems.isEmpty()) {
-            return false;
+            // Note that here, the clonedInvoice is just used as a container object, it doesn't actually represent the final invoice.
+            // For instance, the plugin could return an ITEM_ADJ pointing to an existing invoice: the clonedInvoice object would
+            // not include the items from the original invoice. This doesn't matter anyways, as the dao will create the appropriate entries
+            // and we will re-hydrate the objects before returning (at least in the non dry-run path, see comments in processAccountWithLockAndInputTargetDate).
+            final DefaultInvoice clonedInvoice = (DefaultInvoice) originalInvoice.clone();
```

Comment:
Do we really need to clone -- seems like we remove pretty much everything?

Reply:
I actually went back and forth on this :-) I ended up keeping the `clone()` call to hide the HA complexity (different constructor) but agreed we could go both ways.